server/src/internalClusterTest/java/org/elasticsearch/cluster/coordination/NodeJoiningIT.java

```diff
@@ -0,0 +1,308 @@
+                    blockActionNameOnMockTransportService(
+                        mockTransportService,
+                        List.of(StatefulPreVoteCollector.REQUEST_PRE_VOTE_ACTION_NAME)
+                    );
+                }
+            }
+
+            // We expect the node join request to fail with a FailedToCommitClusterStateException
+            mockLog.addExpectation(
+                new MockLog.SeenEventExpectation(
+                    "failed to commit cluster state",
+                    MasterService.class.getCanonicalName(),
+                    Level.WARN,
+                    "failed to commit cluster state"
+                )
+            );
+
+            /*
+                We expect the cluster to reuse the connection to N and not disconnect it
+                Therefore, this WARN log should not be thrown (#ES-11449)
+             */
+            addJoiningNodeDisconnectedWarnLogFalseExpectation(mockLog);
+
+            // Before we add the new node, assert we haven't changed master nodes yet
+            assertEquals(masterNodeName, internalCluster().getMasterName());
+
+            // Sends a node join request to the original master node. This will fail, and cause a master failover
+            logger.info("Sending node join request");
+            String newNodeName = internalCluster().startDataOnlyNode();
+
+            safeAwait(publishingBanRemovedListener);
+            logger.info("Master publishing ban removed");
+
+            // Wait until the master acknowledges its re-election. The master is only re-elected once it's publishing ban is lifted
+            SubscribableListener<Void> masterKnowsItsReElectedListener = ClusterServiceUtils.addTemporaryStateListener(
+                internalCluster().getInstance(ClusterService.class, masterNodeName),
+                clusterState -> {
+                    DiscoveryNode currentMasterNode = clusterState.nodes().getMasterNode();
+                    long currentTerm = clusterState.coordinationMetadata().term();
+                    return currentMasterNode != null && currentMasterNode.getName().equals(masterNodeName) && currentTerm > originalTerm;
+                }
+            );
+            safeAwait(masterKnowsItsReElectedListener);
+
+            assertEquals(masterNodeName, internalCluster().getMasterName());
+            logger.info("Master has been re-elected");
+
+            try {
+                // Await for N to be in the cluster state of all nodes
+                for (String nodeName : namesOfAllNodesInOriginalCluster) {
+                    ClusterServiceUtils.awaitClusterState(
+                        logger,
+                        clusterState -> nodeExistsWithName(clusterState.nodes(), newNodeName),
+                        internalCluster().clusterService(nodeName)
+                    );
+                }
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
```

Comment:
This should already be true before `startDataOnlyNode` returns, no need to wait. We could check it's true if you'd like, but for that I'd recommend iterating over `internalCluster().getInstances(ClusterService.class)` rather than looking each one up by node name.